sridha2412/assignment2/q1.py

```diff
@@ -33,4 +45,37 @@ def print_ancestors(root, key):
 root.right.right = Node(8)
 root.left.left.left = Node(1)
 root.left.left.right = Node(6)
-print_ancestors(root, 10)
+ancestors = []
+print("Key not found" if not find_ancestors(root, 6, ancestors)
+else ", ".join([str(i) for i in ancestors]))
+
+
+import unittest
+
+class Test(unittest.TestCase):
+    def setUp(self):
+        #       7
+        #      / \
+        #     3   4
+        #    / \   \
+        #   2   5   8
+        #  / \
+        # 1   6
+        self.root = Node(7)
```

Comment:
An alternative to the procedural/imperative style of creating a tree is the constructors of the tree that is driven by the input. In the solution I suggest, leftness or the rightness of the subtrees is decided by the presence of a element in the second or third position of the input tuple. 

Both solution works, the solution that is driven by the data is more flexible as you can generate trees programmatically by simply creating a function that returns the right tuple. In the case of procedural code like below, you would need to generate, not a tuple but generate the code itself. Code generation is fun but it is another complex software problem.

Reply:
I agree, my approach isn't flexible at all. I only did so to quickly test it 😅